Gems/ROS2/Code/Source/Imu/ROS2ImuSensorComponent.cpp

```diff
@@ -77,25 +77,54 @@ namespace ROS2
         required.push_back(AZ_CRC_CE("ROS2Frame"));
     }
 
-    void ROS2ImuSensorComponent::SetupRefreshLoop()
+    void ROS2ImuSensorComponent::Activate()
     {
-        InstallPhysicalCallback();
+        auto ros2Node = ROS2Interface::Get()->GetNode();
+        AZ_Assert(m_sensorConfiguration.m_publishersConfigurations.size() == 1, "Invalid configuration of publishers for IMU sensor");
+        m_imuMsg.header.frame_id = GetFrameID().c_str();
+        const auto publisherConfig = m_sensorConfiguration.m_publishersConfigurations[Internal::kImuMsgType];
+        const auto fullTopic = ROS2Names::GetNamespacedName(GetNamespace(), publisherConfig.m_topic);
+        m_imuPublisher = ros2Node->create_publisher<sensor_msgs::msg::Imu>(fullTopic.data(), publisherConfig.GetQoS());
+
+        m_linearAccelerationCovariance = ToDiagonalCovarianceMatrix(m_imuConfiguration.m_linearAccelerationVariance);
+        m_angularVelocityCovariance = ToDiagonalCovarianceMatrix(m_imuConfiguration.m_angularVelocityVariance);
+        m_orientationCovariance = ToDiagonalCovarianceMatrix(m_imuConfiguration.m_orientationVariance);
+
+        StartSensor(
+            m_sensorConfiguration.m_frequency,
+            [this](float imuDeltaTime, AzPhysics::SceneHandle sceneHandle, float physicsDeltaTime)
+            {
+                if (!m_sensorConfiguration.m_publishingEnabled)
+                {
+                    return;
+                }
+                OnImuEvent(imuDeltaTime, sceneHandle, physicsDeltaTime);
+            },
+            [this]([[maybe_unused]] AzPhysics::SceneHandle sceneHandle, float physicsDeltaTime)
+            {
+                OnPhysicsEvent(sceneHandle);
+            });
     }
 
-    void ROS2ImuSensorComponent::OnPhysicsInitialization(AzPhysics::SceneHandle sceneHandle)
+    void ROS2ImuSensorComponent::Deactivate()
     {
-        AzPhysics::RigidBody* rigidBody = nullptr;
-        AZ::EntityId entityId = GetEntityId();
-        Physics::RigidBodyRequestBus::EventResult(rigidBody, entityId, &Physics::RigidBodyRequests::GetRigidBody);
-        AZ_Assert(rigidBody, "Entity %s does not have rigid body.", entityId.ToString().c_str());
-
-        m_bodyHandle = rigidBody->m_bodyHandle;
+        StopSensor();
+        m_bodyHandle = AzPhysics::InvalidSimulatedBodyHandle;
+        m_imuPublisher.reset();
     }
 
-    void ROS2ImuSensorComponent::OnPhysicsSimulationFinished(AzPhysics::SceneHandle sceneHandle, float deltaTime)
+    void ROS2ImuSensorComponent::OnPhysicsEvent(AzPhysics::SceneHandle sceneHandle)
     {
+        if (m_bodyHandle == AzPhysics::InvalidSimulatedBodyHandle)
+        {
+            AzPhysics::RigidBody* rigidBody = nullptr;
+            AZ::EntityId entityId = GetEntityId();
+            Physics::RigidBodyRequestBus::EventResult(rigidBody, entityId, &Physics::RigidBodyRequests::GetRigidBody);
+            AZ_Assert(rigidBody, "Entity %s does not have rigid body.", entityId.ToString().c_str());
```

Comment:
```suggestion
            AZ_Assert(rigidBody, "Entity %s does not have a rigid body.", entityId.ToString().c_str());
```

Reply:
In release, there will be no assert and it will crash in next line. May we handle null rigid body by logging an error and disconnecting the source instead? We might not want to necessarily crash the entire app because one sensor has an issue.

Reply:
Agree. I changed assert to error + return / stop.